src/stakeinput.cpp

```diff
@@ -8,31 +8,44 @@
 #include "txdb.h"
 #include "wallet/wallet.h"
 
-CPivStake* CPivStake::NewPivStake(const CTxIn& txin)
+static bool HasStakeMinAgeOrDepth(int nHeight, uint32_t nTime, const CBlockIndex* pindex)
+{
+    const Consensus::Params& consensus = Params().GetConsensus();
+    if (consensus.NetworkUpgradeActive(nHeight + 1, Consensus::UPGRADE_ZC_PUBLIC) &&
+            !consensus.HasStakeMinAgeOrDepth(nHeight, nTime, pindex->nHeight, pindex->nTime)) {
+        return error("%s : min age violation - height=%d - time=%d, nHeightBlockFrom=%d, nTimeBlockFrom=%d",
+                     __func__, nHeight, nTime, pindex->nHeight, pindex->nTime);
+    }
+    return true;
+}
+
+CPivStake* CPivStake::NewPivStake(const CTxIn& txin, int nHeight, uint32_t nTime)
 {
     if (txin.IsZerocoinSpend()) {
         error("%s: unable to initialize CPivStake from zerocoin spend", __func__);
         return nullptr;
     }
 
-    // Look for the stake input in the coins cache
+    // Look for the stake input in the coins cache first
     const Coin& coin = pcoinsTip->AccessCoin(txin.prevout);
     if (!coin.IsSpent()) {
-        return new CPivStake(coin.out,
-                             txin.prevout,
-                             mapBlockIndex.at(chainActive[coin.nHeight]->GetBlockHash()));
+        const CBlockIndex* pindexFrom = mapBlockIndex.at(chainActive[coin.nHeight]->GetBlockHash());
+        // Check that the stake has the required depth/age
+        if (!HasStakeMinAgeOrDepth(nHeight, nTime, pindexFrom)) {
+            return nullptr;
+        }
+        // All good
+        return new CPivStake(coin.out, txin.prevout, pindexFrom);
```

Comment:
Nice 3bce1dd6ecdfe5a5c0124f89f71679a6f931a39a
Future thing (or something to do here), will be good to receive a validation state and set each of the possible error causes instead of directly logging the invalidity reason (which is part of the remote disk filling open vector).
It will allow us to add specific test coverage for each of them.

-- I might add it in another PR that I'm cooking if you don't want to do it here --

Reply:
Ya I'd rather leave that for a follow up. This PR has become bigger than expected already.